File: solutions/js/04.js
/**
 * @param {string} code - The code to decipher
 * @returns {string} The deciphered PIN
 */
function decodeSantaPin(code) {
  let result = "";
  let number = 0;

  for (const item of code) {
    switch (item) {
      case "+": {
        const transformed = number + 1;
        number = transformed === 10 ? 0 : transformed;
        break;
      }
      case "-": {
        const transformed = number - 1;
        number = transformed === -1 ? 9 : transformed;
        break;
      }
      case "<":
        number = Number.parseInt(result.at(-1));
        break;
      case "]":
        result += number;
        break;
      default:
        number = Number.parseInt(item);
    }
  }

  return result.length < 4 ? null : result;
}

const decode1 = decodeSantaPin("[1++][2-][3+][<]");
console.log(decode1);
// "3144"

const decode2 = decodeSantaPin("[9+][0-][4][<]");
console.log(decode2);
// "0944"

const decode3 = decodeSantaPin("[1+][2-]");
console.log(decode3);
// null (only 2 digits)

/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 270
  ops: 4037
*/

File: solutions/python/04.py
def decode_santa_pin(code: str) -> str:
    result = ''
    number = 0

    for item in code:
        if (item == "+"):
            transformed = number+1
            number = 0 if transformed == 10 else transformed
        elif (item == "-"):
            transformed = number-1
            number = 9 if transformed == -1 else transformed
        elif (item == "<"):
            number = int(result[-1])
        elif (item == "]"):
            result += str(number)
        elif (item == "["):
            continue
        else:
            number = int(item)

    return result if len(result) == 4 else None


decode1 = decode_santa_pin("[1++][2-][3+][<]")
print(decode1)
# "3144"

decode2 = decode_santa_pin("[9+][0-][4][<]")
print(decode2)
# "0944"

decode3 = decode_santa_pin("[1+][2-]")
print(decode3)
# null (only 2 digits)

"""
  Score: ★★★★★★★ (3 bonus stars for solving it today!)
  Points: 140
  ops: 3498
"""
